lib/config/formatter.go

```diff
@@ -0,0 +1,325 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"regexp"
+	"runtime"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+type textFormatter struct {
+	// ComponentPadding is a padding to pick when displaying
+	// and formatting component field, defaults to DefaultComponentPadding
+	ComponentPadding int
+	// EnableColors enables colored output
+	EnableColors bool
+	// FormatCaller is a function to return (part) of source file path for output.
+	// Defaults to filePathAndLine() if unspecified
+	FormatCaller func() (caller string)
+	// InitFormat is the initial format of log configuration
+	InitFormat []string
+	// InputFormat is the parsed format of each log line
+	InputFormat []string
+}
+
+// DefaultLoggerFormat is the default format in which to display logs (i.e. log.format in teleport.yaml is empty)
+var DefaultLoggerFormat = []string{"level", "component", "message", "caller"}
+var re = regexp.MustCompile(`[a-z]*`)
+
+type writer struct {
+	bytes.Buffer
+}
+
+const (
+	noColor = -1
+	red     = 31
+	yellow  = 33
+	blue    = 36
+	gray    = 37
+)
+
+// CheckAndSetDefaults checks and sets log format configuration
+func (tf *textFormatter) CheckAndSetDefaults() error {
+	// set padding
+	if tf.ComponentPadding == 0 {
+		tf.ComponentPadding = trace.DefaultComponentPadding
+	}
+	// set caller
+	if tf.FormatCaller == nil {
```

Comment:
nit: also I don't think anyone will want to customize this in the initial iteration - let's remove and use `formatCallerWithPathAndLine` directly.